src/sql/finalize.sql

```diff
@@ -154,6 +154,111 @@ IMMUTABLE STRICT PARALLEL SAFE LANGUAGE c AS 'MODULE_PATHNAME', '_vchordg_amhand
 CREATE FUNCTION vchordg_prewarm(regclass) RETURNS TEXT
 STRICT LANGUAGE c AS 'MODULE_PATHNAME', '_vchordg_prewarm_wrapper';
 
+CREATE FUNCTION vchordrq_evaluate_query_recall(
+    query_text text,
+    exact_search boolean default False,
+    gt_epsilon real default 1.9
+)
+RETURNS real
+LANGUAGE plpgsql
+STRICT
+AS $$
+DECLARE
+    query_results vector[];
+    gt_results vector[];
+    match_count integer := 0;
+    gt_count integer;
+    recall real;
+    iter_vector vector;
+    top_k integer;
+    query_probes text;
+    query_epsilon text;
+    gt_probes text;
+    enable_indexscan_original text;
+    enable_bitmapscan_original text;
+    enable_seqscan_original text;
+
+BEGIN
+    enable_indexscan_original := current_setting('enable_indexscan');
+    enable_bitmapscan_original := current_setting('enable_bitmapscan');
+    enable_seqscan_original := current_setting('enable_seqscan');
+    IF NOT exact_search THEN
+        BEGIN
+            query_probes := current_setting('vchordrq.probes');
+            query_epsilon := current_setting('vchordrq.epsilon');
+        EXCEPTION WHEN OTHERS THEN
+            RAISE EXCEPTION 'Could not read vchordrq GUCs. Is the vchordrq extension loaded and configured? Error: %', SQLERRM;
+        END;
+    END IF;
+
+    BEGIN
+        FOR iter_vector IN EXECUTE query_text LOOP
+            query_results := array_append(query_results, iter_vector);
+        END LOOP;
+    EXCEPTION WHEN OTHERS THEN
+        RAISE EXCEPTION 'Error executing ANN query "%": %', query_text, SQLERRM;
+    END;
+
+    top_k := array_length(query_results, 1);
+    IF top_k IS NULL OR top_k = 0 THEN
+        RAISE WARNING 'ANN query returned 0 results. Please check and run the query_text.';
+        RETURN 0.0;
+    END IF;
+
+    IF exact_search THEN
+        SET LOCAL enable_indexscan = off;
+        SET LOCAL enable_bitmapscan = off;
+        SET LOCAL enable_seqscan = on;
+    ELSE
+        IF query_probes = '' THEN
+            gt_probes := '';
+        ELSIF position(',' in query_probes) > 0 THEN
+            gt_probes := '9999,9999';
+        ELSE
+            gt_probes := '9999';
+        END IF;
+        EXECUTE format('SET LOCAL "vchordrq.probes" = %L', gt_probes);
```

Comment:
We may start a subtransaction here.

Reply:
PL/pgSQL doesn't support a transaction inside.

https://stackoverflow.com/questions/14550241/postgresql-cannot-begin-end-transactions-in-pl-pgsql